{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, gzip, shutil\n",
    "\n",
    "'''__get_related_files\n",
    "\n",
    "DESC:\n",
    "    Given some substring, return all files with that substring\n",
    "PARAMS:\n",
    "    files: list of strings of file names\n",
    "    sub: string substring to find in the files\n",
    "OPTIONAL: \n",
    "    not_sub: string a substring that if found in file name don't add\n",
    "'''\n",
    "def __get_related_files(files, sub, not_sub=None):\n",
    "    if not_sub is not None and not_sub != '':\n",
    "        return [x for x in files if sub in x and not_sub not in x]\n",
    "    return [x for x in files if sub in x]\n",
    "\n",
    "'''__make_valid_dir_string\n",
    "\n",
    "DESC:\n",
    "    add / to end of director string if it doesn't have it alread\n",
    "PARAMS:\n",
    "    dir_path: string path to directory\n",
    "RETURNS:\n",
    "    dir path with / at end\n",
    "'''\n",
    "def __make_valid_dir_string(dir_path):\n",
    "    return dir_path + '/' if dir_path[-1] != '/' else dir_path\n",
    "\n",
    "'''__make_dir\n",
    "\n",
    "DESC:\n",
    "    Given a path to directory, check if it exists and if not create it\n",
    "PARAMS:\n",
    "    dir_path: string path to a directory to make or check\n",
    "RETURNS:\n",
    "    None\n",
    "'''\n",
    "def __make_dir(dir_path):\n",
    "    dir_path = __make_valid_dir_string(dir_path)\n",
    "    if not os.path.exists(dir_path): \n",
    "        os.makedirs(dir_path)\n",
    "\n",
    "'''__make_valid_text_file\n",
    "\n",
    "DESC:\n",
    "    make a string into the name for a text file and make sure directory exists\n",
    "PARAMS:\n",
    "    file_name: string a name of the file to save\n",
    "RETURNS:\n",
    "    file name with .txt after\n",
    "'''\n",
    "def __make_valid_text_file(file_name):\n",
    "    file_name = file_name + '.txt' if '.txt' not in file_name else file_name\n",
    "    return file_name\n",
    "\n",
    "'''__make_valid_json_file\n",
    "\n",
    "DESC:\n",
    "    make a string into the name for a text file and make sure the directory exists\n",
    "PARAMS:\n",
    "    file_name: string of the file to save\n",
    "RETURNS:\n",
    "    file name with .json after it\n",
    "'''\n",
    "def __make_valid_json_file(file_name):\n",
    "    file_name = file_name + '.json' if '.json' not in file_name else file_name\n",
    "    return file_name\n",
    "\n",
    "'''__make_valid_csv_file\n",
    "\n",
    "DESC:\n",
    "    make a string into the name for a text file \n",
    "PARAMS:\n",
    "    file_name: string of the file to save\n",
    "RETURNS:\n",
    "    file name with .csv after it\n",
    "'''\n",
    "def __make_valid_csv_file(file_name):\n",
    "    file_name = file_name + '.csv' if '.csv' not in file_name else file_name\n",
    "    return file_name\n",
    "\n",
    "'''__make_valid_fasta_file\n",
    "\n",
    "DESC:\n",
    "    make a string into the name for a text file \n",
    "PARAMS:\n",
    "    file_name: string of the file to save\n",
    "RETURNS:\n",
    "    file name with .fasta after it\n",
    "'''\n",
    "def __make_valid_fasta_file(file_name):\n",
    "    file_name = file_name + '.fasta' if '.fasta' not in file_name else file_name\n",
    "    return file_name\n",
    "\n",
    "'''__file_exists\n",
    "\n",
    "DESC:\n",
    "    find out if a file exists\n",
    "PARAMS:\n",
    "    file_name: string name of the file to check for\n",
    "RETURNS:\n",
    "    bool true if file exists false otherwise\n",
    "'''\n",
    "def __file_exists(file_name):\n",
    "    return os.path.isfile(file_name)\n",
    "\n",
    "'''__gzip\n",
    "\n",
    "DESC:\n",
    "    zip up a file\n",
    "PARAMS: \n",
    "    file_name: str path to the file name to compress\n",
    "OPTIONAL:\n",
    "    delete_old: bool delete the uncompressed file. Default=True\n",
    "RETURNS:\n",
    "    str name of the new compressed file\n",
    "'''\n",
    "def __gzip(file_name, delete_old=True):\n",
    "    compressed_file_name = file_name + '.gz'\n",
    "    with open(file_name, 'rb') as f_in:\n",
    "        with gzip.open(compressed_file_name, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "    delete_old and os.remove(file_name)\n",
    "    return compressed_file_name\n",
    "\n",
    "'''__gunzip\n",
    "\n",
    "DESC:   \n",
    "    unzip a file\n",
    "PARAMS:\n",
    "    compressed_file_name: str name of the compressed file to unzip\n",
    "OPTIONAL:\n",
    "    delete_old: bool delete the compressed file. Default=True\n",
    "RETURNS:\n",
    "    str name of the file unziped\n",
    "'''  \n",
    "def __gunzip(compressed_file_name, delete_old=True):\n",
    "    file_name = compressed_file_name if '.gz' not in compressed_file_name else compressed_file_name.replace('.gz', '')\n",
    "    with gzip.open(compressed_file_name, 'rb') as f_in:\n",
    "        with open(file_name, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "    delete_old and os.remove(compressed_file_name)\n",
    "    return file_name\n",
    "\n",
    "'''__is_gzipped\n",
    "\n",
    "DESC:\n",
    "    determines if a file has been gzipped\n",
    "PARAMS:\n",
    "    file_name: str path to file in question\n",
    "RETURNS:\n",
    "    bool True if file is compressed else False\n",
    "'''\n",
    "def __is_gzipped(file_name):\n",
    "    return '.gz' == file_name[-3:]\n",
    "\n",
    "'''__gzip_dir\n",
    "\n",
    "DESC:\n",
    "    compress a directory with gzip\n",
    "PARAMS:\n",
    "    d: str path to directory\n",
    "OPTIONAL:\n",
    "    delete_old: bool delete the unziped directory. Default=True\n",
    "RETURNS:\n",
    "    path to the new zipped folder\n",
    "'''\n",
    "def __gzip_dir(d, delete_old=True):\n",
    "    root = '/'.join(d.split('/')[:-1])\n",
    "    shutil.make_archive(d, 'zip', root)\n",
    "    delete_old and shutil.rmtree(d)\n",
    "    return d + '.zip'\n",
    "\n",
    "'''__is_json\n",
    "\n",
    "DESC:\n",
    "    determine if a file is a json file based purely on name\n",
    "PARAMS:\n",
    "    file: file to determine if its a json file\n",
    "RETURNS:\n",
    "    bool True if it is a json file False otherwise\n",
    "'''\n",
    "def __is_json(file):\n",
    "    return True if '.json' in file else False\n",
    "\n",
    "'''__is_fasta\n",
    "\n",
    "DESC:\n",
    "    determine if a file is a fasta file based purely on name\n",
    "PARAMS:\n",
    "    file: file to determine if its a fasta file\n",
    "RETURNS:\n",
    "    bool True if it is a fasta file False otherwise\n",
    "'''\n",
    "def __is_fasta(file):\n",
    "    return True if '.fasta' in file else False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def load_experiment(f):\n",
    "    return json.load(open(f, 'r'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spectra generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids={\n",
    "    \"A\":71.037114,\n",
    "    \"R\":156.101111,\n",
    "    \"N\":114.042927,\n",
    "    \"D\":115.026943,\n",
    "    \"C\":103.009185,\n",
    "    \"E\":129.042593,\n",
    "    \"Q\":128.058578,\n",
    "    \"G\":57.021464,\n",
    "    \"H\":137.058912,\n",
    "    \"I\":113.084064,\n",
    "    \"L\":113.084064,\n",
    "    \"K\":128.094963,\n",
    "    \"M\":131.040485,\n",
    "    \"F\":147.068414,\n",
    "    \"P\":97.052764,\n",
    "    \"S\":87.032028,\n",
    "    \"T\":101.047679,\n",
    "    \"U\":150.95363,\n",
    "    \"W\":186.079313,\n",
    "    \"Y\":163.06332,\n",
    "    \"V\":99.068414\n",
    "}\n",
    "\n",
    "'''calc_masses\n",
    "\n",
    "DESC:\n",
    "    calculates the masses/spectrum for a sequence\n",
    "PARAMS:\n",
    "    sequence: str amino acid sequence to change to list of masses\n",
    "    charge: int charge value to calculate masses for\n",
    "RETURNS:\n",
    "    list of floats, float       spectrum and the precursor mass \n",
    "'''\n",
    "def calc_masses(sequence, charge):\n",
    "    masses = []\n",
    "\n",
    "    length = len(sequence)\n",
    "    total = 2 * 1.007825035 + 15.99491463 #This is the mass of water. Adding the mass of water to the sum of all the residue masses gives the mass of the peptide.\n",
    "    for i in range(length):\n",
    "        total +=  amino_acids[sequence[i]]\n",
    "\n",
    "    pre_mz = (total+charge*1.0072764)/charge   \n",
    "\n",
    "    if charge == 1:\n",
    "        #b+\n",
    "        total = 1.007825035 - 0.0005486 #for the H to turn the residue NH on the N-terminus into NH2\n",
    "        for i in range (0, length):\n",
    "            total += amino_acids[sequence[i]]\n",
    "            masses.append(total)\n",
    "            #Since z (the charge) is equal to one, the total here is the m/z\n",
    "\n",
    "        #y+\n",
    "        total = 3 * 1.007825035 + 15.99491463 - 0.0005486 #for the OH to turn the residue CO on the C-terminus into COOH + 1 proton to make NH into NH2 and 1 proton make positively charged\n",
    "        for i in range (0,length):\n",
    "            total += amino_acids[sequence[length-i-1]]\n",
    "            masses.append(total)\n",
    "\n",
    "    elif charge == 2:\n",
    "        #b++\n",
    "        total = 2 * 1.007825035 - 2 * 0.0005486 #adding one more proton this time to make it doubly charged\n",
    "        for i in range (0, length):\n",
    "            total += amino_acids[sequence[i]]\n",
    "            masses.append(total/2)\n",
    "\n",
    "        #y++\n",
    "        total = 4 * 1.007825035 + 15.99491463 - 2 * 0.0005486 #another proton to make doubly charged\n",
    "        for i in range (0, length):\n",
    "            total += amino_acids[sequence[length-i-1]]\n",
    "            masses.append(total/2)\n",
    "        #The masses you get exactly match Spectrum Mill. To get this, I had to make sure to use the mass of H+ and the mass of H when appropriate.\n",
    "\n",
    "    return masses, pre_mz\n",
    "\n",
    "'''gen_spectra\n",
    "\n",
    "DESC:\n",
    "    generates mass spectra for sequences\n",
    "PARAMS:\n",
    "    sequences: list of strings sequences to generate spectra for\n",
    "RETURNS:\n",
    "    list of dictionaries of the form {'spectrum': list of floats, 'precursor_mass': float}\n",
    "'''\n",
    "def gen_spectra(sequenences):\n",
    "    spectra = []\n",
    "    for sequence in sequenences:\n",
    "        this_entry = {}\n",
    "        mass_1, _ = calc_masses(sequence, 1)\n",
    "        mass_2, pre_mz = calc_masses(sequence, 2)\n",
    "        this_spectra = mass_1 + mass_2\n",
    "        this_spectra.sort()\n",
    "        this_entry['spectrum'] = this_spectra\n",
    "        this_entry['precursor_mass'] = pre_mz\n",
    "        spectra.append(this_entry)\n",
    "\n",
    "    return spectra"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spectra writing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n"
     ]
    }
   ],
   "source": [
    "from pyopenms import MSExperiment, MSSpectrum, MzMLFile, Peak1D, Precursor\n",
    "\n",
    "'''write_mzml\n",
    "\n",
    "DESC:\n",
    "    create a mass spectrum file in mzml of sequences\n",
    "PARAMS:\n",
    "    file_name: str name to save the file in\n",
    "    spectra: list of dictionaries of the form [{spectrum: list[floats], precursor_mass: float, sequence: str}]\n",
    "             This data is written to file (the spectrum and the precursor)\n",
    "OPTIONAL:\n",
    "    title_prefix: str name to give as prefix to the name of each spectrum. Default=Spectrum\n",
    "    output_dir: str name of the directory to save files to. Default=./\n",
    "    compress: bool whether or not to compress the file. Comrpesses with gzip. Default=True\n",
    "RETURNS:\n",
    "    list of strings of file paths\n",
    "'''\n",
    "def write_mzml(file_name, spectra, title_prefix='Spectrum ', output_dir='./', compress=True):\n",
    "    if '.mzml' not in file_name.lower():\n",
    "        file_name += '.mzML'\n",
    "    output_dir = __make_valid_dir_string(output_dir)\n",
    "    __make_dir(output_dir)\n",
    "    output_file = output_dir + file_name\n",
    "\n",
    "    exp = MSExperiment()\n",
    "    sp_count = 0\n",
    "    for spectrum in spectra:\n",
    "        spec = MSSpectrum()\n",
    "        spec.setMSLevel(2)\n",
    "        name = str.encode(title_prefix + str(sp_count))\n",
    "        spec.setName(name)\n",
    "        sp_count += 1\n",
    "        \n",
    "        i = [500 for _ in spectrum['spectrum']]\n",
    "        spec.set_peaks([spectrum['spectrum'], i])\n",
    "        spec.setMSLevel(2)\n",
    "        prec = Precursor()\n",
    "        prec.setCharge(2)\n",
    "        prec.setMZ(spectrum['precursor_mass'])\n",
    "        spec.setPrecursors([prec])\n",
    "        spec.sortByPosition()\n",
    "        exp.addSpectrum(spec)\n",
    "\n",
    "    MzMLFile().store(output_file, exp)\n",
    "    output_file = __gzip(output_file) if compress else output_file\n",
    "\n",
    "    return output_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Crux scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subprocess import call\n",
    "\n",
    "crux_to_rm = ['tide-search.decoy.txt', 'tide-search.log.txt', 'tide-search.params.txt']\n",
    "\n",
    "def __parse_spectrum_name(spec_name):\n",
    "    return str(spec_name.split('/')[-1]).lower().replace('.mzml', '')\n",
    "\n",
    "def __parse_db_name(db_name):\n",
    "    return  str(db_name.split('/')[-1]).replace('.fasta', '')\n",
    "\n",
    "def __index_db_files(path_to_crux_cmd, db_files):\n",
    "    idx_names = []\n",
    "    num_dbs = len(db_files)\n",
    "\n",
    "    for i, db_file in enumerate(db_files):\n",
    "        print('On database {}/{}[{}%]\\r'.format(i+1, num_dbs, int(((i+1)/num_dbs) * 100)), end=\"\")\n",
    "        \n",
    "        this_output_dir = '/'.join(str(db_file).split('/')[:-1])\n",
    "        this_output_dir = __make_valid_dir_string(this_output_dir) + 'indexed/'\n",
    "        __make_dir(this_output_dir)\n",
    "        idx_name = this_output_dir + str(db_file).replace('.fasta', '_index').split('/')[-1]\n",
    "\n",
    "        indx_cmd = [\n",
    "            path_to_crux_cmd, \n",
    "            'tide-index', \n",
    "            db_file, \n",
    "            idx_name, \n",
    "            '--min-length', '2', \n",
    "            '--min-mass', '50', \n",
    "            '--output-dir', this_output_dir, \n",
    "            '--overwrite', 'T', \n",
    "            '--min-peaks', '2', \n",
    "            '--precursor-window', '1000000000',\n",
    "            '--enzyme', 'no-enzyme', \n",
    "            '--verbosity', '0'\n",
    "        ]\n",
    "        call(indx_cmd)\n",
    "        # remove extra output files\n",
    "        os.remove(this_output_dir + 'tide-index.params.txt')\n",
    "        os.remove(this_output_dir + 'tide-index.log.txt')\n",
    "\n",
    "        idx_names.append(idx_name)\n",
    "    return idx_names\n",
    "\n",
    "def __remove_indices(index_file):\n",
    "    if isinstance(index_file, list):\n",
    "        index_file = index_file[0]\n",
    "    rm_dir = '/'.join(index_file.split('/')[:-1])\n",
    "    shutil.rmtree(rm_dir)\n",
    "\n",
    "\n",
    "'''crux_search\n",
    "\n",
    "DESC:\n",
    "    use the crux tool to score spectra against databases\n",
    "PARAMS:\n",
    "    spectra_files: list of str paths to all the spectra (.mzML) files\n",
    "    database_files: list of str paths to all the database (.fasta) files\n",
    "    path_to_crux_cmd: str path to the executable for crux.\n",
    "    output_dir: str path to the directory to save files\n",
    "OPTIONAL:\n",
    "    compress: bool compress the output result. Default=True\n",
    "RETURNS:\n",
    "    list of str of output files\n",
    "'''\n",
    "def __crux_search(spectra_files, database_files, path_to_crux_cmd, output_dir, compress=True):\n",
    "    output_dir = __make_valid_dir_string(output_dir) + 'search_output/'\n",
    "    __make_dir(output_dir)\n",
    "    spec_dir = '/'.join(spectra_files[0].split('/')[:-1])\n",
    "\n",
    "    is_compressed = __is_gzipped(spectra_files[0])\n",
    "    print('Pre-indexing database files...')\n",
    "    indexed_db_files = __index_db_files(path_to_crux_cmd, database_files)\n",
    "    print('\\nDone. Scoring..')\n",
    "\n",
    "    output_count = 0\n",
    "    output_files = []\n",
    "    num_dbs = len(indexed_db_files)\n",
    "    num_specs = len(spectra_files)\n",
    "\n",
    "    for i, spec_file in enumerate(spectra_files):\n",
    "        spec_file = spec_file if not is_compressed else __gunzip(spec_file)\n",
    "        for j, database_file in enumerate(indexed_db_files):\n",
    "            this_db_name = __parse_db_name(database_file)\n",
    "            print('On spectrum {}/{}[{}%]\\tOn database file {}/{}[{}%]\\r'.format(i+1, num_specs, int(((i+1)/num_specs) * 100), j+1, num_dbs, int(((j+1)/num_dbs)*100)), end=\"\")\n",
    "            this_output_dir = output_dir + '{}_vs_{}'.format(__parse_spectrum_name(spec_file), this_db_name)\n",
    "            search_cmd = [\n",
    "                path_to_crux_cmd, \n",
    "                'tide-search', \n",
    "                spec_file, \n",
    "                database_file, \n",
    "                '--min-length', '2', \n",
    "                '--min-mass', '50', \n",
    "                '--output-dir', this_output_dir, \n",
    "                '--overwrite', 'T', \n",
    "                '--min-peaks', '2', \n",
    "                '--precursor-window', '1000000000',\n",
    "                '--enzyme', 'no-enzyme', \n",
    "                '--verbosity', '0'\n",
    "                ] \n",
    "            call(search_cmd)\n",
    "            output_count += 1\n",
    "            o = this_output_dir + '/tide-search.target.txt' if not compress else __gzip(this_output_dir + '/tide-search.target.txt')\n",
    "            o_tsv = o.replace('.txt', '.tsv')\n",
    "            os.rename(o, o_tsv)\n",
    "            output_files.append(o_tsv)\n",
    "\n",
    "            # saving space, so should remove the extra stuff\n",
    "            if is_compressed:\n",
    "                for rm in crux_to_rm:\n",
    "                    os.remove(this_output_dir + '/' + rm)\n",
    "\n",
    "        # if the files were compressed, we were trying to save disk space so just remove the mzml files\n",
    "        # is_compressed and os.remove(spec_file)\n",
    "\n",
    "    # is_compressed and os.rmdir(spec_dir)\n",
    "    __remove_indices(indexed_db_files)\n",
    "    return output_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load crux search results\n",
    "\n",
    "### NOTE: the (int) after a protein name is the start index of the sequence, 1 based not 0, so subtract 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "desired_cols = ['file', 'xcorr score', 'sequence', 'protein id', 'target/decoy', 'xcorr rank']\n",
    "\n",
    "def load_search_result(f):\n",
    "    if __is_gzipped(f):\n",
    "        f = __gunzip(f)\n",
    "    df = pd.read_csv(f, '\\t', header=0)\n",
    "    return df[desired_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add hybrid proteins to fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_hyb_prots(fasta_file, prots):\n",
    "    with open(fasta_file, 'a') as o:\n",
    "        o.write('\\n')\n",
    "        c = 0\n",
    "        for p in prots:\n",
    "            if 'hybrid' in p['name'].lower():\n",
    "                o.write('>sp|idsomething{}|{}\\n'.format(c, p['name']))\n",
    "                o.write(p['protein'] + '\\n')\n",
    "                c += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "# START RUNNING FROM HERE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "import copy\n",
    "\n",
    "experiment_file = '/Users/zacharymcgrath/Desktop/Experiment output/small prots 1000 peps 5 hybs zscoresum crux scoring func/experiment_data.json'\n",
    "save_dir = __make_valid_dir_string('/Users/zacharymcgrath/Desktop/Experiment output/small prots 1000 peps 5 hybs crux run')\n",
    "\n",
    "exp = load_experiment(experiment_file)\n",
    "peptides = exp['experiment_info']['peptides']\n",
    "keyed_peps = {}\n",
    "for pep in peptides:\n",
    "    keyed_peps[pep['peptide_name']] = copy.deepcopy(pep)\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate spectra files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating spectra 1004/1005[99%]\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "spec_files = []\n",
    "\n",
    "spec_save_dir = __make_valid_dir_string(__make_valid_dir_string(save_dir) + 'spectra')\n",
    "__make_dir(spec_save_dir)\n",
    "\n",
    "pep_c = 0\n",
    "pep_num = len(peptides)\n",
    "for pep in peptides:\n",
    "    print('Generating spectra {}/{}[{}%]\\r'.format(pep_c, pep_num, int( (float(pep_c)/(pep_num)) * 100)), end=\"\")\n",
    "    pep_c += 1\n",
    "    \n",
    "    name = pep['peptide_name']\n",
    "    seq = pep['peptide_sequence']\n",
    "    spectra = gen_spectra([seq])\n",
    "    spec_files.append(write_mzml(name, spectra, output_dir=spec_save_dir))\n",
    "print('\\nDone')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run crux search on spectra files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pre-indexing database files...\n",
      "On database 1/1[100%]\n",
      "Done. Scoring..\n",
      "On spectrum 1005/1005[100%]\tOn database file 1/1[100%]\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "fasta_file = '/Users/zacharymcgrath/Documents/Layer_Research/Proteomics_Experiments/Database_Experiments/test-protein.fasta'\n",
    "path_to_crux_cmd = '/Users/zacharymcgrath/Documents/Layer_Research/crux/bin/crux'\n",
    "\n",
    "# add hybrid proteins to file before we run search\n",
    "add_hyb_prots(fasta_file, exp['experiment_info']['proteins'])\n",
    "\n",
    "score_files = __crux_search(spec_files, [fasta_file], path_to_crux_cmd, save_dir)\n",
    "print('\\nDone')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Next steps\n",
    "\n",
    "Ok by this point we have files with the scores for every peptide against the one database. We can get their names now and reconstruct this. We need to do the following:\n",
    "1. find out which protein crux said it was from \n",
    "2. Determine if it was the the proper position\n",
    "3. If more than 1 candidate, then filter that out\n",
    "4. construct a list of scores for a protein/peptide pair to see if it properly scored or not. For situations where there was no peptide generated, rank = 0. If the correct position or protein was not found, give a score of -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "dfs = []\n",
    "\n",
    "for sf in score_files:\n",
    "    dfs.append(load_search_result(sf))\n",
    "    \n",
    "df = pd.concat(dfs, ignore_index=True, sort =False)\n",
    "print('Done')\n",
    "# print(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize search results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "search_results = {}\n",
    "\n",
    "def take_name(s):\n",
    "        \n",
    "    rm_num = lambda x: x.split('(')[0]\n",
    "    rm_bar = lambda x: x.split('|')[2]\n",
    "    \n",
    "    return [rm_num(rm_bar(y)) for y in s.split(',')]\n",
    "        \n",
    "\n",
    "for p in peptides:\n",
    "    name = p['peptide_name']\n",
    "    search_results[name] = []\n",
    "    \n",
    "    take_start_pos = lambda s: int(s.split('(')[1].split(')')[0])\n",
    "    for i, row in df.iterrows():\n",
    "        if name in row['file'] and row['target/decoy'] != 'decoy':\n",
    "            search_results[name].append({\n",
    "                'protein': take_name(row['protein id']),\n",
    "                'starting_position': take_start_pos(row['protein id']),\n",
    "                'sequence': row['sequence'],\n",
    "                'peptide_name': name,\n",
    "                'rank': row['xcorr rank']\n",
    "            })\n",
    "print('Done')\n",
    "# print(search_results)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize by parent protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "ps = exp['experiment_info']['proteins']\n",
    "prots = {}\n",
    "for p in ps:\n",
    "    prots[p['name']] = p['sequence'] if 'sequence' in p else p['protein']\n",
    "\n",
    "sorted_results = {}\n",
    "for p in prots: \n",
    "    sorted_results[p] = []\n",
    "\n",
    "for peptide in search_results:\n",
    "    for r in search_results[peptide]:\n",
    "        for prot in r['protein']:\n",
    "            sorted_results[prot].append(r)\n",
    "print('Done')\n",
    "# print(sorted_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compress results\n",
    "For each protein, we want to know if the peptide associated with that result was scored correctly or not, and update some sort of array. This list should be the length of the protein sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On protein 0/9[0%]\r",
      "On protein 1/9[11%]\r",
      "On protein 2/9[22%]\r",
      "On protein 3/9[33%]\r",
      "On protein 4/9[44%]\r",
      "On protein 5/9[55%]\r",
      "On protein 6/9[66%]\r",
      "On protein 7/9[77%]\r",
      "On protein 8/9[88%]\r\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "# wrong_parent = 'r'\n",
    "# wrong_loc = 'y'\n",
    "hit = 'g'\n",
    "no_data = 'r'\n",
    "\n",
    "protein_peptide_data = {}\n",
    "\n",
    "prot_c = 0\n",
    "prot_num = len(prots)\n",
    "for prot in prots:\n",
    "    print('On protein {}/{}[{}%]\\r'.format(prot_c, prot_num, int( (float(prot_c)/(prot_num)) * 100)), end=\"\")\n",
    "    prot_c += 1\n",
    "    \n",
    "    prot_data = [[] for _ in range(len(prots[prot]))]\n",
    "    for r in sorted_results[prot]:\n",
    "        # get the REAL peptide data\n",
    "        info = keyed_peps[r['peptide_name']]\n",
    "        \n",
    "        if prot in r['protein']:\n",
    "            result_start_idx = int(r['starting_position']) -1\n",
    "            # we have the correct parent protein\n",
    "            if prot == info['parent_name']:\n",
    "                # we have the right starting position\n",
    "                if int(info['start_index']) == result_start_idx: #-1 becuase they use 1 based counting\n",
    "                    prot_data[info['start_index']].append((hit, int(r['rank']), r['peptide_name']))\n",
    "                    \n",
    "# THESE ARE COMMENTED OUT TO REDUCE THE NUMBER OF DATA POINTS WE SHOW\n",
    "#                 else:\n",
    "#                     prot_data[result_start_idx].append((wrong_loc, int(r['rank'])))\n",
    "                    \n",
    "#             # we don't have the correct parent protein\n",
    "#             else:\n",
    "#                 prot_data[result_start_idx].append((wrong_parent, int(r['rank'])))\n",
    "                \n",
    "    protein_peptide_data[prot] = prot_data\n",
    "print('\\nDone')\n",
    "#print(protein_peptide_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peptide cleanup\n",
    "\n",
    "go through every peptide and check the correct parent protein at the correct position and see if it managed to find it. If not give it the no data and a -1. Reduce the number of points. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On peptide 0/1005[0%]\r",
      "On peptide 1/1005[0%]\r",
      "On peptide 2/1005[0%]\r",
      "On peptide 3/1005[0%]\r",
      "On peptide 4/1005[0%]\r",
      "On peptide 5/1005[0%]\r",
      "On peptide 6/1005[0%]\r",
      "On peptide 7/1005[0%]\r",
      "On peptide 8/1005[0%]\r",
      "On peptide 9/1005[0%]\r",
      "On peptide 10/1005[0%]\r",
      "On peptide 11/1005[1%]\r",
      "On peptide 12/1005[1%]\r",
      "On peptide 13/1005[1%]\r",
      "On peptide 14/1005[1%]\r",
      "On peptide 15/1005[1%]\r",
      "On peptide 16/1005[1%]\r",
      "On peptide 17/1005[1%]\r",
      "On peptide 18/1005[1%]\r",
      "On peptide 19/1005[1%]\r",
      "On peptide 20/1005[1%]\r",
      "On peptide 21/1005[2%]\r",
      "On peptide 22/1005[2%]\r",
      "On peptide 23/1005[2%]\r",
      "On peptide 24/1005[2%]\r",
      "On peptide 25/1005[2%]\r",
      "On peptide 26/1005[2%]\r",
      "On peptide 27/1005[2%]\r",
      "On peptide 28/1005[2%]\r",
      "On peptide 29/1005[2%]\r",
      "On peptide 30/1005[2%]\r",
      "On peptide 31/1005[3%]\r",
      "On peptide 32/1005[3%]\r",
      "On peptide 33/1005[3%]\r",
      "On peptide 34/1005[3%]\r",
      "On peptide 35/1005[3%]\r",
      "On peptide 36/1005[3%]\r",
      "On peptide 37/1005[3%]\r",
      "On peptide 38/1005[3%]\r",
      "On peptide 39/1005[3%]\r",
      "On peptide 40/1005[3%]\r",
      "On peptide 41/1005[4%]\r",
      "On peptide 42/1005[4%]\r",
      "On peptide 43/1005[4%]\r",
      "On peptide 44/1005[4%]\r",
      "On peptide 45/1005[4%]\r",
      "On peptide 46/1005[4%]\r",
      "On peptide 47/1005[4%]\r",
      "On peptide 48/1005[4%]\r",
      "On peptide 49/1005[4%]\r",
      "On peptide 50/1005[4%]\r",
      "On peptide 51/1005[5%]\r",
      "On peptide 52/1005[5%]\r",
      "On peptide 53/1005[5%]\r",
      "On peptide 54/1005[5%]\r",
      "On peptide 55/1005[5%]\r",
      "On peptide 56/1005[5%]\r",
      "On peptide 57/1005[5%]\r",
      "On peptide 58/1005[5%]\r",
      "On peptide 59/1005[5%]\r",
      "On peptide 60/1005[5%]\r",
      "On peptide 61/1005[6%]\r",
      "On peptide 62/1005[6%]\r",
      "On peptide 63/1005[6%]\r",
      "On peptide 64/1005[6%]\r",
      "On peptide 65/1005[6%]\r",
      "On peptide 66/1005[6%]\r",
      "On peptide 67/1005[6%]\r",
      "On peptide 68/1005[6%]\r",
      "On peptide 69/1005[6%]\r",
      "On peptide 70/1005[6%]\r",
      "On peptide 71/1005[7%]\r",
      "On peptide 72/1005[7%]\r",
      "On peptide 73/1005[7%]\r",
      "On peptide 74/1005[7%]\r",
      "On peptide 75/1005[7%]\r",
      "On peptide 76/1005[7%]\r",
      "On peptide 77/1005[7%]\r",
      "On peptide 78/1005[7%]\r",
      "On peptide 79/1005[7%]\r",
      "On peptide 80/1005[7%]\r",
      "On peptide 81/1005[8%]\r",
      "On peptide 82/1005[8%]\r",
      "On peptide 83/1005[8%]\r",
      "On peptide 84/1005[8%]\r",
      "On peptide 85/1005[8%]\r",
      "On peptide 86/1005[8%]\r",
      "On peptide 87/1005[8%]\r",
      "On peptide 88/1005[8%]\r",
      "On peptide 89/1005[8%]\r",
      "On peptide 90/1005[8%]\r",
      "On peptide 91/1005[9%]\r",
      "On peptide 92/1005[9%]\r",
      "On peptide 93/1005[9%]\r",
      "On peptide 94/1005[9%]\r",
      "On peptide 95/1005[9%]\r",
      "On peptide 96/1005[9%]\r",
      "On peptide 97/1005[9%]\r",
      "On peptide 98/1005[9%]\r",
      "On peptide 99/1005[9%]\r",
      "On peptide 100/1005[9%]\r",
      "On peptide 101/1005[10%]\r",
      "On peptide 102/1005[10%]\r",
      "On peptide 103/1005[10%]\r",
      "On peptide 104/1005[10%]\r",
      "On peptide 105/1005[10%]\r",
      "On peptide 106/1005[10%]\r",
      "On peptide 107/1005[10%]\r",
      "On peptide 108/1005[10%]\r",
      "On peptide 109/1005[10%]\r",
      "On peptide 110/1005[10%]\r",
      "On peptide 111/1005[11%]\r",
      "On peptide 112/1005[11%]\r",
      "On peptide 113/1005[11%]\r",
      "On peptide 114/1005[11%]\r",
      "On peptide 115/1005[11%]\r",
      "On peptide 116/1005[11%]\r",
      "On peptide 117/1005[11%]\r",
      "On peptide 118/1005[11%]\r",
      "On peptide 119/1005[11%]\r",
      "On peptide 120/1005[11%]\r",
      "On peptide 121/1005[12%]\r",
      "On peptide 122/1005[12%]\r",
      "On peptide 123/1005[12%]\r",
      "On peptide 124/1005[12%]\r",
      "On peptide 125/1005[12%]\r",
      "On peptide 126/1005[12%]\r",
      "On peptide 127/1005[12%]\r",
      "On peptide 128/1005[12%]\r",
      "On peptide 129/1005[12%]\r",
      "On peptide 130/1005[12%]\r",
      "On peptide 131/1005[13%]\r",
      "On peptide 132/1005[13%]\r",
      "On peptide 133/1005[13%]\r",
      "On peptide 134/1005[13%]\r",
      "On peptide 135/1005[13%]\r",
      "On peptide 136/1005[13%]\r",
      "On peptide 137/1005[13%]\r",
      "On peptide 138/1005[13%]\r",
      "On peptide 139/1005[13%]\r",
      "On peptide 140/1005[13%]\r",
      "On peptide 141/1005[14%]\r",
      "On peptide 142/1005[14%]\r",
      "On peptide 143/1005[14%]\r",
      "On peptide 144/1005[14%]\r",
      "On peptide 145/1005[14%]\r",
      "On peptide 146/1005[14%]\r",
      "On peptide 147/1005[14%]\r",
      "On peptide 148/1005[14%]\r",
      "On peptide 149/1005[14%]\r",
      "On peptide 150/1005[14%]\r",
      "On peptide 151/1005[15%]\r",
      "On peptide 152/1005[15%]\r",
      "On peptide 153/1005[15%]\r",
      "On peptide 154/1005[15%]\r",
      "On peptide 155/1005[15%]\r",
      "On peptide 156/1005[15%]\r",
      "On peptide 157/1005[15%]\r",
      "On peptide 158/1005[15%]\r",
      "On peptide 159/1005[15%]\r",
      "On peptide 160/1005[15%]\r",
      "On peptide 161/1005[16%]\r",
      "On peptide 162/1005[16%]\r",
      "On peptide 163/1005[16%]\r",
      "On peptide 164/1005[16%]\r",
      "On peptide 165/1005[16%]\r",
      "On peptide 166/1005[16%]\r",
      "On peptide 167/1005[16%]\r",
      "On peptide 168/1005[16%]\r",
      "On peptide 169/1005[16%]\r",
      "On peptide 170/1005[16%]\r",
      "On peptide 171/1005[17%]\r",
      "On peptide 172/1005[17%]\r",
      "On peptide 173/1005[17%]\r",
      "On peptide 174/1005[17%]\r",
      "On peptide 175/1005[17%]\r",
      "On peptide 176/1005[17%]\r",
      "On peptide 177/1005[17%]\r",
      "On peptide 178/1005[17%]\r",
      "On peptide 179/1005[17%]\r",
      "On peptide 180/1005[17%]\r",
      "On peptide 181/1005[18%]\r",
      "On peptide 182/1005[18%]\r",
      "On peptide 183/1005[18%]\r",
      "On peptide 184/1005[18%]\r",
      "On peptide 185/1005[18%]\r",
      "On peptide 186/1005[18%]\r",
      "On peptide 187/1005[18%]\r",
      "On peptide 188/1005[18%]\r",
      "On peptide 189/1005[18%]\r",
      "On peptide 190/1005[18%]\r",
      "On peptide 191/1005[19%]\r",
      "On peptide 192/1005[19%]\r",
      "On peptide 193/1005[19%]\r",
      "On peptide 194/1005[19%]\r",
      "On peptide 195/1005[19%]\r",
      "On peptide 196/1005[19%]\r",
      "On peptide 197/1005[19%]\r",
      "On peptide 198/1005[19%]\r",
      "On peptide 199/1005[19%]\r",
      "On peptide 200/1005[19%]\r",
      "On peptide 201/1005[20%]\r",
      "On peptide 202/1005[20%]\r",
      "On peptide 203/1005[20%]\r",
      "On peptide 204/1005[20%]\r",
      "On peptide 205/1005[20%]\r",
      "On peptide 206/1005[20%]\r",
      "On peptide 207/1005[20%]\r",
      "On peptide 208/1005[20%]\r",
      "On peptide 209/1005[20%]\r",
      "On peptide 210/1005[20%]\r",
      "On peptide 211/1005[20%]\r",
      "On peptide 212/1005[21%]\r",
      "On peptide 213/1005[21%]\r",
      "On peptide 214/1005[21%]\r",
      "On peptide 215/1005[21%]\r",
      "On peptide 216/1005[21%]\r",
      "On peptide 217/1005[21%]\r",
      "On peptide 218/1005[21%]\r",
      "On peptide 219/1005[21%]\r",
      "On peptide 220/1005[21%]\r",
      "On peptide 221/1005[21%]\r",
      "On peptide 222/1005[22%]\r",
      "On peptide 223/1005[22%]\r",
      "On peptide 224/1005[22%]\r",
      "On peptide 225/1005[22%]\r",
      "On peptide 226/1005[22%]\r",
      "On peptide 227/1005[22%]\r",
      "On peptide 228/1005[22%]\r",
      "On peptide 229/1005[22%]\r",
      "On peptide 230/1005[22%]\r",
      "On peptide 231/1005[22%]\r",
      "On peptide 232/1005[23%]\r",
      "On peptide 233/1005[23%]\r",
      "On peptide 234/1005[23%]\r",
      "On peptide 235/1005[23%]\r",
      "On peptide 236/1005[23%]\r",
      "On peptide 237/1005[23%]\r",
      "On peptide 238/1005[23%]\r",
      "On peptide 239/1005[23%]\r",
      "On peptide 240/1005[23%]\r",
      "On peptide 241/1005[23%]\r",
      "On peptide 242/1005[24%]\r",
      "On peptide 243/1005[24%]\r",
      "On peptide 244/1005[24%]\r",
      "On peptide 245/1005[24%]\r",
      "On peptide 246/1005[24%]\r",
      "On peptide 247/1005[24%]\r",
      "On peptide 248/1005[24%]\r",
      "On peptide 249/1005[24%]\r",
      "On peptide 250/1005[24%]\r",
      "On peptide 251/1005[24%]\r",
      "On peptide 252/1005[25%]\r",
      "On peptide 253/1005[25%]\r",
      "On peptide 254/1005[25%]\r",
      "On peptide 255/1005[25%]\r",
      "On peptide 256/1005[25%]\r",
      "On peptide 257/1005[25%]\r",
      "On peptide 258/1005[25%]\r",
      "On peptide 259/1005[25%]\r",
      "On peptide 260/1005[25%]\r",
      "On peptide 261/1005[25%]\r",
      "On peptide 262/1005[26%]\r",
      "On peptide 263/1005[26%]\r",
      "On peptide 264/1005[26%]\r",
      "On peptide 265/1005[26%]\r",
      "On peptide 266/1005[26%]\r",
      "On peptide 267/1005[26%]\r",
      "On peptide 268/1005[26%]\r",
      "On peptide 269/1005[26%]\r",
      "On peptide 270/1005[26%]\r",
      "On peptide 271/1005[26%]\r",
      "On peptide 272/1005[27%]\r",
      "On peptide 273/1005[27%]\r",
      "On peptide 274/1005[27%]\r",
      "On peptide 275/1005[27%]\r",
      "On peptide 276/1005[27%]\r",
      "On peptide 277/1005[27%]\r",
      "On peptide 278/1005[27%]\r",
      "On peptide 279/1005[27%]\r",
      "On peptide 280/1005[27%]\r",
      "On peptide 281/1005[27%]\r",
      "On peptide 282/1005[28%]\r",
      "On peptide 283/1005[28%]\r",
      "On peptide 284/1005[28%]\r",
      "On peptide 285/1005[28%]\r",
      "On peptide 286/1005[28%]\r",
      "On peptide 287/1005[28%]\r",
      "On peptide 288/1005[28%]\r",
      "On peptide 289/1005[28%]\r",
      "On peptide 290/1005[28%]\r",
      "On peptide 291/1005[28%]\r",
      "On peptide 292/1005[29%]\r",
      "On peptide 293/1005[29%]\r",
      "On peptide 294/1005[29%]\r",
      "On peptide 295/1005[29%]\r",
      "On peptide 296/1005[29%]\r",
      "On peptide 297/1005[29%]\r",
      "On peptide 298/1005[29%]\r",
      "On peptide 299/1005[29%]\r",
      "On peptide 300/1005[29%]\r",
      "On peptide 301/1005[29%]\r",
      "On peptide 302/1005[30%]\r",
      "On peptide 303/1005[30%]\r",
      "On peptide 304/1005[30%]\r",
      "On peptide 305/1005[30%]\r",
      "On peptide 306/1005[30%]\r",
      "On peptide 307/1005[30%]\r",
      "On peptide 308/1005[30%]\r",
      "On peptide 309/1005[30%]\r",
      "On peptide 310/1005[30%]\r",
      "On peptide 311/1005[30%]\r",
      "On peptide 312/1005[31%]\r",
      "On peptide 313/1005[31%]\r",
      "On peptide 314/1005[31%]\r",
      "On peptide 315/1005[31%]\r",
      "On peptide 316/1005[31%]\r",
      "On peptide 317/1005[31%]\r",
      "On peptide 318/1005[31%]\r",
      "On peptide 319/1005[31%]\r",
      "On peptide 320/1005[31%]\r",
      "On peptide 321/1005[31%]\r",
      "On peptide 322/1005[32%]\r",
      "On peptide 323/1005[32%]\r",
      "On peptide 324/1005[32%]\r",
      "On peptide 325/1005[32%]\r",
      "On peptide 326/1005[32%]\r",
      "On peptide 327/1005[32%]\r",
      "On peptide 328/1005[32%]\r",
      "On peptide 329/1005[32%]\r",
      "On peptide 330/1005[32%]\r",
      "On peptide 331/1005[32%]\r",
      "On peptide 332/1005[33%]\r",
      "On peptide 333/1005[33%]\r",
      "On peptide 334/1005[33%]\r",
      "On peptide 335/1005[33%]\r",
      "On peptide 336/1005[33%]\r",
      "On peptide 337/1005[33%]\r",
      "On peptide 338/1005[33%]\r",
      "On peptide 339/1005[33%]\r",
      "On peptide 340/1005[33%]\r",
      "On peptide 341/1005[33%]\r",
      "On peptide 342/1005[34%]\r",
      "On peptide 343/1005[34%]\r",
      "On peptide 344/1005[34%]\r",
      "On peptide 345/1005[34%]\r",
      "On peptide 346/1005[34%]\r",
      "On peptide 347/1005[34%]\r",
      "On peptide 348/1005[34%]\r",
      "On peptide 349/1005[34%]\r",
      "On peptide 350/1005[34%]\r",
      "On peptide 351/1005[34%]\r",
      "On peptide 352/1005[35%]\r",
      "On peptide 353/1005[35%]\r",
      "On peptide 354/1005[35%]\r",
      "On peptide 355/1005[35%]\r",
      "On peptide 356/1005[35%]\r",
      "On peptide 357/1005[35%]\r",
      "On peptide 358/1005[35%]\r",
      "On peptide 359/1005[35%]\r",
      "On peptide 360/1005[35%]\r",
      "On peptide 361/1005[35%]\r",
      "On peptide 362/1005[36%]\r",
      "On peptide 363/1005[36%]\r",
      "On peptide 364/1005[36%]\r",
      "On peptide 365/1005[36%]\r",
      "On peptide 366/1005[36%]\r",
      "On peptide 367/1005[36%]\r",
      "On peptide 368/1005[36%]\r",
      "On peptide 369/1005[36%]\r",
      "On peptide 370/1005[36%]\r",
      "On peptide 371/1005[36%]\r",
      "On peptide 372/1005[37%]\r",
      "On peptide 373/1005[37%]\r",
      "On peptide 374/1005[37%]\r",
      "On peptide 375/1005[37%]\r",
      "On peptide 376/1005[37%]\r",
      "On peptide 377/1005[37%]\r",
      "On peptide 378/1005[37%]\r",
      "On peptide 379/1005[37%]\r",
      "On peptide 380/1005[37%]\r",
      "On peptide 381/1005[37%]\r",
      "On peptide 382/1005[38%]\r",
      "On peptide 383/1005[38%]\r",
      "On peptide 384/1005[38%]\r",
      "On peptide 385/1005[38%]\r",
      "On peptide 386/1005[38%]\r",
      "On peptide 387/1005[38%]\r",
      "On peptide 388/1005[38%]\r",
      "On peptide 389/1005[38%]\r",
      "On peptide 390/1005[38%]\r",
      "On peptide 391/1005[38%]\r",
      "On peptide 392/1005[39%]\r",
      "On peptide 393/1005[39%]\r",
      "On peptide 394/1005[39%]\r",
      "On peptide 395/1005[39%]\r",
      "On peptide 396/1005[39%]\r",
      "On peptide 397/1005[39%]\r",
      "On peptide 398/1005[39%]\r",
      "On peptide 399/1005[39%]\r",
      "On peptide 400/1005[39%]\r",
      "On peptide 401/1005[39%]\r",
      "On peptide 402/1005[40%]\r",
      "On peptide 403/1005[40%]\r",
      "On peptide 404/1005[40%]\r",
      "On peptide 405/1005[40%]\r",
      "On peptide 406/1005[40%]\r",
      "On peptide 407/1005[40%]\r",
      "On peptide 408/1005[40%]\r",
      "On peptide 409/1005[40%]\r",
      "On peptide 410/1005[40%]\r",
      "On peptide 411/1005[40%]\r",
      "On peptide 412/1005[40%]\r",
      "On peptide 413/1005[41%]\r",
      "On peptide 414/1005[41%]\r",
      "On peptide 415/1005[41%]\r",
      "On peptide 416/1005[41%]\r",
      "On peptide 417/1005[41%]\r",
      "On peptide 418/1005[41%]\r",
      "On peptide 419/1005[41%]\r",
      "On peptide 420/1005[41%]\r",
      "On peptide 421/1005[41%]\r",
      "On peptide 422/1005[41%]\r",
      "On peptide 423/1005[42%]\r",
      "On peptide 424/1005[42%]\r",
      "On peptide 425/1005[42%]\r",
      "On peptide 426/1005[42%]\r",
      "On peptide 427/1005[42%]\r",
      "On peptide 428/1005[42%]\r",
      "On peptide 429/1005[42%]\r",
      "On peptide 430/1005[42%]\r",
      "On peptide 431/1005[42%]\r",
      "On peptide 432/1005[42%]\r",
      "On peptide 433/1005[43%]\r",
      "On peptide 434/1005[43%]\r",
      "On peptide 435/1005[43%]\r",
      "On peptide 436/1005[43%]\r",
      "On peptide 437/1005[43%]\r",
      "On peptide 438/1005[43%]\r",
      "On peptide 439/1005[43%]\r",
      "On peptide 440/1005[43%]\r",
      "On peptide 441/1005[43%]\r",
      "On peptide 442/1005[43%]\r",
      "On peptide 443/1005[44%]\r",
      "On peptide 444/1005[44%]\r",
      "On peptide 445/1005[44%]\r",
      "On peptide 446/1005[44%]\r",
      "On peptide 447/1005[44%]\r",
      "On peptide 448/1005[44%]\r",
      "On peptide 449/1005[44%]\r",
      "On peptide 450/1005[44%]\r",
      "On peptide 451/1005[44%]\r",
      "On peptide 452/1005[44%]\r",
      "On peptide 453/1005[45%]\r",
      "On peptide 454/1005[45%]\r",
      "On peptide 455/1005[45%]\r",
      "On peptide 456/1005[45%]\r",
      "On peptide 457/1005[45%]\r",
      "On peptide 458/1005[45%]\r",
      "On peptide 459/1005[45%]\r",
      "On peptide 460/1005[45%]\r",
      "On peptide 461/1005[45%]\r",
      "On peptide 462/1005[45%]\r",
      "On peptide 463/1005[46%]\r",
      "On peptide 464/1005[46%]\r",
      "On peptide 465/1005[46%]\r",
      "On peptide 466/1005[46%]\r",
      "On peptide 467/1005[46%]\r",
      "On peptide 468/1005[46%]\r",
      "On peptide 469/1005[46%]\r",
      "On peptide 470/1005[46%]\r",
      "On peptide 471/1005[46%]\r",
      "On peptide 472/1005[46%]\r",
      "On peptide 473/1005[47%]\r",
      "On peptide 474/1005[47%]\r",
      "On peptide 475/1005[47%]\r",
      "On peptide 476/1005[47%]\r",
      "On peptide 477/1005[47%]\r",
      "On peptide 478/1005[47%]\r",
      "On peptide 479/1005[47%]\r",
      "On peptide 480/1005[47%]\r",
      "On peptide 481/1005[47%]\r",
      "On peptide 482/1005[47%]\r",
      "On peptide 483/1005[48%]\r",
      "On peptide 484/1005[48%]\r",
      "On peptide 485/1005[48%]\r",
      "On peptide 486/1005[48%]\r",
      "On peptide 487/1005[48%]\r",
      "On peptide 488/1005[48%]\r",
      "On peptide 489/1005[48%]\r",
      "On peptide 490/1005[48%]\r",
      "On peptide 491/1005[48%]\r",
      "On peptide 492/1005[48%]\r",
      "On peptide 493/1005[49%]\r",
      "On peptide 494/1005[49%]\r",
      "On peptide 495/1005[49%]\r",
      "On peptide 496/1005[49%]\r",
      "On peptide 497/1005[49%]\r",
      "On peptide 498/1005[49%]\r",
      "On peptide 499/1005[49%]\r",
      "On peptide 500/1005[49%]\r",
      "On peptide 501/1005[49%]\r",
      "On peptide 502/1005[49%]\r",
      "On peptide 503/1005[50%]\r",
      "On peptide 504/1005[50%]\r",
      "On peptide 505/1005[50%]\r",
      "On peptide 506/1005[50%]\r",
      "On peptide 507/1005[50%]\r",
      "On peptide 508/1005[50%]\r",
      "On peptide 509/1005[50%]\r",
      "On peptide 510/1005[50%]\r",
      "On peptide 511/1005[50%]\r",
      "On peptide 512/1005[50%]\r",
      "On peptide 513/1005[51%]\r",
      "On peptide 514/1005[51%]\r",
      "On peptide 515/1005[51%]\r",
      "On peptide 516/1005[51%]\r",
      "On peptide 517/1005[51%]\r",
      "On peptide 518/1005[51%]\r",
      "On peptide 519/1005[51%]\r",
      "On peptide 520/1005[51%]\r",
      "On peptide 521/1005[51%]\r",
      "On peptide 522/1005[51%]\r",
      "On peptide 523/1005[52%]\r",
      "On peptide 524/1005[52%]\r",
      "On peptide 525/1005[52%]\r",
      "On peptide 526/1005[52%]\r",
      "On peptide 527/1005[52%]\r",
      "On peptide 528/1005[52%]\r",
      "On peptide 529/1005[52%]\r",
      "On peptide 530/1005[52%]\r",
      "On peptide 531/1005[52%]\r",
      "On peptide 532/1005[52%]\r",
      "On peptide 533/1005[53%]\r",
      "On peptide 534/1005[53%]\r",
      "On peptide 535/1005[53%]\r",
      "On peptide 536/1005[53%]\r",
      "On peptide 537/1005[53%]\r",
      "On peptide 538/1005[53%]\r",
      "On peptide 539/1005[53%]\r",
      "On peptide 540/1005[53%]\r",
      "On peptide 541/1005[53%]\r",
      "On peptide 542/1005[53%]\r",
      "On peptide 543/1005[54%]\r",
      "On peptide 544/1005[54%]\r",
      "On peptide 545/1005[54%]\r",
      "On peptide 546/1005[54%]\r",
      "On peptide 547/1005[54%]\r",
      "On peptide 548/1005[54%]\r",
      "On peptide 549/1005[54%]\r",
      "On peptide 550/1005[54%]\r",
      "On peptide 551/1005[54%]\r",
      "On peptide 552/1005[54%]\r",
      "On peptide 553/1005[55%]\r",
      "On peptide 554/1005[55%]\r",
      "On peptide 555/1005[55%]\r",
      "On peptide 556/1005[55%]\r",
      "On peptide 557/1005[55%]\r",
      "On peptide 558/1005[55%]\r",
      "On peptide 559/1005[55%]\r",
      "On peptide 560/1005[55%]\r",
      "On peptide 561/1005[55%]\r",
      "On peptide 562/1005[55%]\r",
      "On peptide 563/1005[56%]\r",
      "On peptide 564/1005[56%]\r",
      "On peptide 565/1005[56%]\r",
      "On peptide 566/1005[56%]\r",
      "On peptide 567/1005[56%]\r",
      "On peptide 568/1005[56%]\r",
      "On peptide 569/1005[56%]\r",
      "On peptide 570/1005[56%]\r",
      "On peptide 571/1005[56%]\r",
      "On peptide 572/1005[56%]\r",
      "On peptide 573/1005[57%]\r",
      "On peptide 574/1005[57%]\r",
      "On peptide 575/1005[57%]\r",
      "On peptide 576/1005[57%]\r",
      "On peptide 577/1005[57%]\r",
      "On peptide 578/1005[57%]\r",
      "On peptide 579/1005[57%]\r",
      "On peptide 580/1005[57%]\r",
      "On peptide 581/1005[57%]\r",
      "On peptide 582/1005[57%]\r",
      "On peptide 583/1005[58%]\r",
      "On peptide 584/1005[58%]\r",
      "On peptide 585/1005[58%]\r",
      "On peptide 586/1005[58%]\r",
      "On peptide 587/1005[58%]\r",
      "On peptide 588/1005[58%]\r",
      "On peptide 589/1005[58%]\r",
      "On peptide 590/1005[58%]\r",
      "On peptide 591/1005[58%]\r",
      "On peptide 592/1005[58%]\r",
      "On peptide 593/1005[59%]\r",
      "On peptide 594/1005[59%]\r",
      "On peptide 595/1005[59%]\r",
      "On peptide 596/1005[59%]\r",
      "On peptide 597/1005[59%]\r",
      "On peptide 598/1005[59%]\r",
      "On peptide 599/1005[59%]\r",
      "On peptide 600/1005[59%]\r",
      "On peptide 601/1005[59%]\r",
      "On peptide 602/1005[59%]\r",
      "On peptide 603/1005[60%]\r",
      "On peptide 604/1005[60%]\r",
      "On peptide 605/1005[60%]\r",
      "On peptide 606/1005[60%]\r",
      "On peptide 607/1005[60%]\r",
      "On peptide 608/1005[60%]\r",
      "On peptide 609/1005[60%]\r",
      "On peptide 610/1005[60%]\r",
      "On peptide 611/1005[60%]\r",
      "On peptide 612/1005[60%]\r",
      "On peptide 613/1005[60%]\r",
      "On peptide 614/1005[61%]\r",
      "On peptide 615/1005[61%]\r",
      "On peptide 616/1005[61%]\r",
      "On peptide 617/1005[61%]\r",
      "On peptide 618/1005[61%]\r",
      "On peptide 619/1005[61%]\r",
      "On peptide 620/1005[61%]\r",
      "On peptide 621/1005[61%]\r",
      "On peptide 622/1005[61%]\r",
      "On peptide 623/1005[61%]\r",
      "On peptide 624/1005[62%]\r",
      "On peptide 625/1005[62%]\r",
      "On peptide 626/1005[62%]\r",
      "On peptide 627/1005[62%]\r",
      "On peptide 628/1005[62%]\r",
      "On peptide 629/1005[62%]\r",
      "On peptide 630/1005[62%]\r",
      "On peptide 631/1005[62%]\r",
      "On peptide 632/1005[62%]\r",
      "On peptide 633/1005[62%]\r",
      "On peptide 634/1005[63%]\r",
      "On peptide 635/1005[63%]\r",
      "On peptide 636/1005[63%]\r",
      "On peptide 637/1005[63%]\r",
      "On peptide 638/1005[63%]\r",
      "On peptide 639/1005[63%]\r",
      "On peptide 640/1005[63%]\r",
      "On peptide 641/1005[63%]\r",
      "On peptide 642/1005[63%]\r",
      "On peptide 643/1005[63%]\r",
      "On peptide 644/1005[64%]\r",
      "On peptide 645/1005[64%]\r",
      "On peptide 646/1005[64%]\r",
      "On peptide 647/1005[64%]\r",
      "On peptide 648/1005[64%]\r",
      "On peptide 649/1005[64%]\r",
      "On peptide 650/1005[64%]\r",
      "On peptide 651/1005[64%]\r",
      "On peptide 652/1005[64%]\r",
      "On peptide 653/1005[64%]\r",
      "On peptide 654/1005[65%]\r",
      "On peptide 655/1005[65%]\r",
      "On peptide 656/1005[65%]\r",
      "On peptide 657/1005[65%]\r",
      "On peptide 658/1005[65%]\r",
      "On peptide 659/1005[65%]\r",
      "On peptide 660/1005[65%]\r",
      "On peptide 661/1005[65%]\r",
      "On peptide 662/1005[65%]\r",
      "On peptide 663/1005[65%]\r",
      "On peptide 664/1005[66%]\r",
      "On peptide 665/1005[66%]\r",
      "On peptide 666/1005[66%]\r",
      "On peptide 667/1005[66%]\r",
      "On peptide 668/1005[66%]\r",
      "On peptide 669/1005[66%]\r",
      "On peptide 670/1005[66%]\r",
      "On peptide 671/1005[66%]\r",
      "On peptide 672/1005[66%]\r",
      "On peptide 673/1005[66%]\r",
      "On peptide 674/1005[67%]\r",
      "On peptide 675/1005[67%]\r",
      "On peptide 676/1005[67%]\r",
      "On peptide 677/1005[67%]\r",
      "On peptide 678/1005[67%]\r",
      "On peptide 679/1005[67%]\r",
      "On peptide 680/1005[67%]\r",
      "On peptide 681/1005[67%]\r",
      "On peptide 682/1005[67%]\r",
      "On peptide 683/1005[67%]\r",
      "On peptide 684/1005[68%]\r",
      "On peptide 685/1005[68%]\r",
      "On peptide 686/1005[68%]\r",
      "On peptide 687/1005[68%]\r",
      "On peptide 688/1005[68%]\r",
      "On peptide 689/1005[68%]\r",
      "On peptide 690/1005[68%]\r",
      "On peptide 691/1005[68%]\r",
      "On peptide 692/1005[68%]\r",
      "On peptide 693/1005[68%]\r",
      "On peptide 694/1005[69%]\r",
      "On peptide 695/1005[69%]\r",
      "On peptide 696/1005[69%]\r",
      "On peptide 697/1005[69%]\r",
      "On peptide 698/1005[69%]\r",
      "On peptide 699/1005[69%]\r",
      "On peptide 700/1005[69%]\r",
      "On peptide 701/1005[69%]\r",
      "On peptide 702/1005[69%]\r",
      "On peptide 703/1005[69%]\r",
      "On peptide 704/1005[70%]\r",
      "On peptide 705/1005[70%]\r",
      "On peptide 706/1005[70%]\r",
      "On peptide 707/1005[70%]\r",
      "On peptide 708/1005[70%]\r",
      "On peptide 709/1005[70%]\r",
      "On peptide 710/1005[70%]\r",
      "On peptide 711/1005[70%]\r",
      "On peptide 712/1005[70%]\r",
      "On peptide 713/1005[70%]\r",
      "On peptide 714/1005[71%]\r",
      "On peptide 715/1005[71%]\r",
      "On peptide 716/1005[71%]\r",
      "On peptide 717/1005[71%]\r",
      "On peptide 718/1005[71%]\r",
      "On peptide 719/1005[71%]\r",
      "On peptide 720/1005[71%]\r",
      "On peptide 721/1005[71%]\r",
      "On peptide 722/1005[71%]\r",
      "On peptide 723/1005[71%]\r",
      "On peptide 724/1005[72%]\r",
      "On peptide 725/1005[72%]\r",
      "On peptide 726/1005[72%]\r",
      "On peptide 727/1005[72%]\r",
      "On peptide 728/1005[72%]\r",
      "On peptide 729/1005[72%]\r",
      "On peptide 730/1005[72%]\r",
      "On peptide 731/1005[72%]\r",
      "On peptide 732/1005[72%]\r",
      "On peptide 733/1005[72%]\r",
      "On peptide 734/1005[73%]\r",
      "On peptide 735/1005[73%]\r",
      "On peptide 736/1005[73%]\r",
      "On peptide 737/1005[73%]\r",
      "On peptide 738/1005[73%]\r",
      "On peptide 739/1005[73%]\r",
      "On peptide 740/1005[73%]\r",
      "On peptide 741/1005[73%]\r",
      "On peptide 742/1005[73%]\r",
      "On peptide 743/1005[73%]\r",
      "On peptide 744/1005[74%]\r",
      "On peptide 745/1005[74%]\r",
      "On peptide 746/1005[74%]\r",
      "On peptide 747/1005[74%]\r",
      "On peptide 748/1005[74%]\r",
      "On peptide 749/1005[74%]\r",
      "On peptide 750/1005[74%]\r",
      "On peptide 751/1005[74%]\r",
      "On peptide 752/1005[74%]\r",
      "On peptide 753/1005[74%]\r",
      "On peptide 754/1005[75%]\r",
      "On peptide 755/1005[75%]\r",
      "On peptide 756/1005[75%]\r",
      "On peptide 757/1005[75%]\r",
      "On peptide 758/1005[75%]\r",
      "On peptide 759/1005[75%]\r",
      "On peptide 760/1005[75%]\r",
      "On peptide 761/1005[75%]\r",
      "On peptide 762/1005[75%]\r",
      "On peptide 763/1005[75%]\r",
      "On peptide 764/1005[76%]\r",
      "On peptide 765/1005[76%]\r",
      "On peptide 766/1005[76%]\r",
      "On peptide 767/1005[76%]\r",
      "On peptide 768/1005[76%]\r",
      "On peptide 769/1005[76%]\r",
      "On peptide 770/1005[76%]\r",
      "On peptide 771/1005[76%]\r",
      "On peptide 772/1005[76%]\r",
      "On peptide 773/1005[76%]\r",
      "On peptide 774/1005[77%]\r",
      "On peptide 775/1005[77%]\r",
      "On peptide 776/1005[77%]\r",
      "On peptide 777/1005[77%]\r",
      "On peptide 778/1005[77%]\r",
      "On peptide 779/1005[77%]\r",
      "On peptide 780/1005[77%]\r",
      "On peptide 781/1005[77%]\r",
      "On peptide 782/1005[77%]\r",
      "On peptide 783/1005[77%]\r",
      "On peptide 784/1005[78%]\r",
      "On peptide 785/1005[78%]\r",
      "On peptide 786/1005[78%]\r",
      "On peptide 787/1005[78%]\r",
      "On peptide 788/1005[78%]\r",
      "On peptide 789/1005[78%]\r",
      "On peptide 790/1005[78%]\r",
      "On peptide 791/1005[78%]\r",
      "On peptide 792/1005[78%]\r",
      "On peptide 793/1005[78%]\r",
      "On peptide 794/1005[79%]\r",
      "On peptide 795/1005[79%]\r",
      "On peptide 796/1005[79%]\r",
      "On peptide 797/1005[79%]\r",
      "On peptide 798/1005[79%]\r",
      "On peptide 799/1005[79%]\r",
      "On peptide 800/1005[79%]\r",
      "On peptide 801/1005[79%]\r",
      "On peptide 802/1005[79%]\r",
      "On peptide 803/1005[79%]\r",
      "On peptide 804/1005[80%]\r",
      "On peptide 805/1005[80%]\r",
      "On peptide 806/1005[80%]\r",
      "On peptide 807/1005[80%]\r",
      "On peptide 808/1005[80%]\r",
      "On peptide 809/1005[80%]\r",
      "On peptide 810/1005[80%]\r",
      "On peptide 811/1005[80%]\r",
      "On peptide 812/1005[80%]\r",
      "On peptide 813/1005[80%]\r",
      "On peptide 814/1005[80%]\r",
      "On peptide 815/1005[81%]\r",
      "On peptide 816/1005[81%]\r",
      "On peptide 817/1005[81%]\r",
      "On peptide 818/1005[81%]\r",
      "On peptide 819/1005[81%]\r",
      "On peptide 820/1005[81%]\r",
      "On peptide 821/1005[81%]\r",
      "On peptide 822/1005[81%]\r",
      "On peptide 823/1005[81%]\r",
      "On peptide 824/1005[81%]\r",
      "On peptide 825/1005[82%]\r",
      "On peptide 826/1005[82%]\r",
      "On peptide 827/1005[82%]\r",
      "On peptide 828/1005[82%]\r",
      "On peptide 829/1005[82%]\r",
      "On peptide 830/1005[82%]\r",
      "On peptide 831/1005[82%]\r",
      "On peptide 832/1005[82%]\r",
      "On peptide 833/1005[82%]\r",
      "On peptide 834/1005[82%]\r",
      "On peptide 835/1005[83%]\r",
      "On peptide 836/1005[83%]\r",
      "On peptide 837/1005[83%]\r",
      "On peptide 838/1005[83%]\r",
      "On peptide 839/1005[83%]\r",
      "On peptide 840/1005[83%]\r",
      "On peptide 841/1005[83%]\r",
      "On peptide 842/1005[83%]\r",
      "On peptide 843/1005[83%]\r",
      "On peptide 844/1005[83%]\r",
      "On peptide 845/1005[84%]\r",
      "On peptide 846/1005[84%]\r",
      "On peptide 847/1005[84%]\r",
      "On peptide 848/1005[84%]\r",
      "On peptide 849/1005[84%]\r",
      "On peptide 850/1005[84%]\r",
      "On peptide 851/1005[84%]\r",
      "On peptide 852/1005[84%]\r",
      "On peptide 853/1005[84%]\r",
      "On peptide 854/1005[84%]\r",
      "On peptide 855/1005[85%]\r",
      "On peptide 856/1005[85%]\r",
      "On peptide 857/1005[85%]\r",
      "On peptide 858/1005[85%]\r",
      "On peptide 859/1005[85%]\r",
      "On peptide 860/1005[85%]\r",
      "On peptide 861/1005[85%]\r",
      "On peptide 862/1005[85%]\r",
      "On peptide 863/1005[85%]\r",
      "On peptide 864/1005[85%]\r",
      "On peptide 865/1005[86%]\r",
      "On peptide 866/1005[86%]\r",
      "On peptide 867/1005[86%]\r",
      "On peptide 868/1005[86%]\r",
      "On peptide 869/1005[86%]\r",
      "On peptide 870/1005[86%]\r",
      "On peptide 871/1005[86%]\r",
      "On peptide 872/1005[86%]\r",
      "On peptide 873/1005[86%]\r",
      "On peptide 874/1005[86%]\r",
      "On peptide 875/1005[87%]\r",
      "On peptide 876/1005[87%]\r",
      "On peptide 877/1005[87%]\r",
      "On peptide 878/1005[87%]\r",
      "On peptide 879/1005[87%]\r",
      "On peptide 880/1005[87%]\r",
      "On peptide 881/1005[87%]\r",
      "On peptide 882/1005[87%]\r",
      "On peptide 883/1005[87%]\r",
      "On peptide 884/1005[87%]\r",
      "On peptide 885/1005[88%]\r",
      "On peptide 886/1005[88%]\r",
      "On peptide 887/1005[88%]\r",
      "On peptide 888/1005[88%]\r",
      "On peptide 889/1005[88%]\r",
      "On peptide 890/1005[88%]\r",
      "On peptide 891/1005[88%]\r",
      "On peptide 892/1005[88%]\r",
      "On peptide 893/1005[88%]\r",
      "On peptide 894/1005[88%]\r",
      "On peptide 895/1005[89%]\r",
      "On peptide 896/1005[89%]\r",
      "On peptide 897/1005[89%]\r",
      "On peptide 898/1005[89%]\r",
      "On peptide 899/1005[89%]\r",
      "On peptide 900/1005[89%]\r",
      "On peptide 901/1005[89%]\r",
      "On peptide 902/1005[89%]\r",
      "On peptide 903/1005[89%]\r",
      "On peptide 904/1005[89%]\r",
      "On peptide 905/1005[90%]\r",
      "On peptide 906/1005[90%]\r",
      "On peptide 907/1005[90%]\r",
      "On peptide 908/1005[90%]\r",
      "On peptide 909/1005[90%]\r",
      "On peptide 910/1005[90%]\r",
      "On peptide 911/1005[90%]\r",
      "On peptide 912/1005[90%]\r",
      "On peptide 913/1005[90%]\r",
      "On peptide 914/1005[90%]\r",
      "On peptide 915/1005[91%]\r",
      "On peptide 916/1005[91%]\r",
      "On peptide 917/1005[91%]\r",
      "On peptide 918/1005[91%]\r",
      "On peptide 919/1005[91%]\r",
      "On peptide 920/1005[91%]\r",
      "On peptide 921/1005[91%]\r",
      "On peptide 922/1005[91%]\r",
      "On peptide 923/1005[91%]\r",
      "On peptide 924/1005[91%]\r",
      "On peptide 925/1005[92%]\r",
      "On peptide 926/1005[92%]\r",
      "On peptide 927/1005[92%]\r",
      "On peptide 928/1005[92%]\r",
      "On peptide 929/1005[92%]\r",
      "On peptide 930/1005[92%]\r",
      "On peptide 931/1005[92%]\r",
      "On peptide 932/1005[92%]\r",
      "On peptide 933/1005[92%]\r",
      "On peptide 934/1005[92%]\r",
      "On peptide 935/1005[93%]\r",
      "On peptide 936/1005[93%]\r",
      "On peptide 937/1005[93%]\r",
      "On peptide 938/1005[93%]\r",
      "On peptide 939/1005[93%]\r",
      "On peptide 940/1005[93%]\r",
      "On peptide 941/1005[93%]\r",
      "On peptide 942/1005[93%]\r",
      "On peptide 943/1005[93%]\r",
      "On peptide 944/1005[93%]\r",
      "On peptide 945/1005[94%]\r",
      "On peptide 946/1005[94%]\r",
      "On peptide 947/1005[94%]\r",
      "On peptide 948/1005[94%]\r",
      "On peptide 949/1005[94%]\r",
      "On peptide 950/1005[94%]\r",
      "On peptide 951/1005[94%]\r",
      "On peptide 952/1005[94%]\r",
      "On peptide 953/1005[94%]\r",
      "On peptide 954/1005[94%]\r",
      "On peptide 955/1005[95%]\r",
      "On peptide 956/1005[95%]\r",
      "On peptide 957/1005[95%]\r",
      "On peptide 958/1005[95%]\r",
      "On peptide 959/1005[95%]\r",
      "On peptide 960/1005[95%]\r",
      "On peptide 961/1005[95%]\r",
      "On peptide 962/1005[95%]\r",
      "On peptide 963/1005[95%]\r",
      "On peptide 964/1005[95%]\r",
      "On peptide 965/1005[96%]\r",
      "On peptide 966/1005[96%]\r",
      "On peptide 967/1005[96%]\r",
      "On peptide 968/1005[96%]\r",
      "On peptide 969/1005[96%]\r",
      "On peptide 970/1005[96%]\r",
      "On peptide 971/1005[96%]\r",
      "On peptide 972/1005[96%]\r",
      "On peptide 973/1005[96%]\r",
      "On peptide 974/1005[96%]\r",
      "On peptide 975/1005[97%]\r",
      "On peptide 976/1005[97%]\r",
      "On peptide 977/1005[97%]\r",
      "On peptide 978/1005[97%]\r",
      "On peptide 979/1005[97%]\r",
      "On peptide 980/1005[97%]\r",
      "On peptide 981/1005[97%]\r",
      "On peptide 982/1005[97%]\r",
      "On peptide 983/1005[97%]\r",
      "On peptide 984/1005[97%]\r",
      "On peptide 985/1005[98%]\r",
      "On peptide 986/1005[98%]\r",
      "On peptide 987/1005[98%]\r",
      "On peptide 988/1005[98%]\r",
      "On peptide 989/1005[98%]\r",
      "On peptide 990/1005[98%]\r",
      "On peptide 991/1005[98%]\r",
      "On peptide 992/1005[98%]\r",
      "On peptide 993/1005[98%]\r",
      "On peptide 994/1005[98%]\r",
      "On peptide 995/1005[99%]\r",
      "On peptide 996/1005[99%]\r",
      "On peptide 997/1005[99%]\r",
      "On peptide 998/1005[99%]\r",
      "On peptide 999/1005[99%]\r",
      "On peptide 1000/1005[99%]\r",
      "On peptide 1001/1005[99%]\r",
      "On peptide 1002/1005[99%]\r",
      "On peptide 1003/1005[99%]\r",
      "On peptide 1004/1005[99%]\r\n",
      "37/1005 peptides did not get correct hit\n"
     ]
    }
   ],
   "source": [
    "def has_correct(prot_data, pos):\n",
    "    rs = prot_data[pos]\n",
    "    return any([r[0] == hit for r in rs])\n",
    "\n",
    "number_missed = 0\n",
    "pep_c = 0\n",
    "pep_num = len(keyed_peps)\n",
    "for pep_name, pep_info in keyed_peps.items():\n",
    "    print('On peptide {}/{}[{}%]\\r'.format(pep_c, pep_num, int( (float(pep_c)/(pep_num)) * 100)), end=\"\")\n",
    "    pep_c += 1\n",
    "    \n",
    "    pos = pep_info['start_index']\n",
    "    corr_parent = pep_info['parent_name']\n",
    "    # if we don't have a point in the right starting position, add a point at it saying its wrong\n",
    "    if not has_correct(protein_peptide_data[corr_parent], pos):\n",
    "        protein_peptide_data[corr_parent][pos].append((no_data, -1))\n",
    "        number_missed += 1\n",
    "    # remove any duplicates from this peptide and take only the highest rank\n",
    "    else:\n",
    "        pep_points = [x for x in protein_peptide_data[corr_parent][pos] if x[2] == pep_name]\n",
    "        pep_points.sort(key=lambda x: x[1])\n",
    "        to_remove = [] if len(pep_points) < 2 else pep_points[1:]\n",
    "        for rm in to_remove:\n",
    "            protein_peptide_data[corr_parent][pos].remove(rm)\n",
    "        \n",
    "print('\\n{}/{} peptides did not get correct hit'.format(number_missed, len(keyed_peps)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot these bad bois"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On protein 9/9[100%]\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import gridspec\n",
    "from math import ceil\n",
    "from statistics import mean\n",
    "\n",
    "max_aas = 75\n",
    "plot_width = 14\n",
    "\n",
    "legend_els = [\n",
    "    Line2D([0], [0], color='r', lw=4, label='wrong parent'),\n",
    "    #Line2D([0], [0], color='y', lw=4, label='right parent wrong location'),\n",
    "    Line2D([0], [0], color='g', lw=4, label='correct')\n",
    "    #Line2D([0], [0], color='c', lw=4, label='correct hit missing')\n",
    "]\n",
    "\n",
    "prot_plot_save_dir = __make_valid_dir_string(__make_valid_dir_string(save_dir) + 'protein_plots/')\n",
    "__make_dir(prot_plot_save_dir)\n",
    "\n",
    "prot_c = 0\n",
    "prot_num = len(protein_peptide_data)\n",
    "for prot in protein_peptide_data:\n",
    "    prot_c += 1\n",
    "    print('On protein {}/{}[{}%]\\r'.format(prot_c, prot_num, int( (float(prot_c)/(prot_num)) * 100)), end=\"\")\n",
    "    num_subplots = ceil(len(prots[prot]) / max_aas)\n",
    "    fig, ax = plt.subplots(num_subplots, figsize=(plot_width, 2.5 * num_subplots))\n",
    "    \n",
    "    \n",
    "    for i in range(num_subplots):\n",
    "        this_axis = ax[i]\n",
    "        # get the data needed for this row\n",
    "        end = min([((i+1) * max_aas), len(prots[prot])])\n",
    "        this_data = protein_peptide_data[prot][i*max_aas:end]\n",
    "        this_label = prots[prot][i*max_aas:end]\n",
    "        for x in range(len(this_data)):\n",
    "            if len(this_data[x]) > 0:\n",
    "                # boxplot this\n",
    "                to_box_plot = [x[1] for x in this_data[x]]\n",
    "                color = 'r' if mean(to_box_plot) < 0 else 'g'\n",
    "                bplot = this_axis.boxplot(to_box_plot, positions=[x])\n",
    "                # color it\n",
    "                for attr in ['boxes', 'whiskers', 'caps', 'fliers', 'medians']:\n",
    "                    for item in bplot[attr]:\n",
    "                        item.set(color=color)\n",
    "                \n",
    "#                 for points in this_data[x]:\n",
    "#                     this_axis.plot([x], [points[1]], marker='o', color=points[0])\n",
    "                    \n",
    "        this_axis.set_xticks([j for j in range(len(this_data))])\n",
    "        this_axis.set_xticklabels(this_label)\n",
    "        this_axis.set_yticks([-2, -1, 0, 1, 2, 3, 4, 5, 6])\n",
    "        this_axis.set_yticklabels(['', 'not found', '0', '1', '2', '3', '4', '5', ''])\n",
    "        y_label = 'sequence:  {} - {}'.format(i * max_aas, end - 1)\n",
    "        this_axis.set_ylabel(y_label, fontsize=8)\n",
    "        this_axis.legend(handles=legend_els, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    fig.suptitle(prot)\n",
    "    plot_save_name = prot_plot_save_dir + prot\n",
    "    fig.savefig(plot_save_name)\n",
    "    #fig.show()\n",
    "    plt.close()\n",
    "print('\\nDone')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
